packages/base/matrix-room.gts

```diff
@@ -0,0 +1,415 @@
+import {
+  contains,
+  containsMany,
+  field,
+  Component,
+  Card,
+  primitive,
+  CardBase,
+  createFromSerialized,
+} from './card-api';
+import StringCard from './string';
+import DateTimeCard from './datetime';
+import IntegerCard from './integer';
+import MarkdownCard from './markdown';
+import { BoxelMessage } from '@cardstack/boxel-ui';
+import cssVar from '@cardstack/boxel-ui/helpers/css-var';
+import { formatRFC3339 } from 'date-fns';
+import Modifier from 'ember-modifier';
+import { type LooseSingleCardDocument } from '@cardstack/runtime-common';
+
+const roomMembers = new Map<string, RoomMemberCard>();
+
+// this is so we can have triple equals equivalent room member cards
+function upsertRoomMember(
+  userId: string,
+  displayName?: string
+): RoomMemberCard {
+  let member = roomMembers.get(userId);
+  if (!member) {
+    member = new RoomMemberCard({ userId });
+    roomMembers.set(userId, member);
+  }
+
+  // patch in the display name in case we don't have one yet
+  if (displayName && member.displayName === undefined) {
+    member.displayName = displayName;
+  }
+
+  return member;
+}
+
+class JSONView extends Component<typeof MatrixEventCard> {
+  <template>
+    <pre>{{this.json}}</pre>
+  </template>
+
+  get json() {
+    return JSON.stringify(this.args.model, null, 2);
+  }
+}
+
+class MatrixEventCard extends CardBase {
+  static [primitive]: MatrixEvent;
+  static embedded = class Embedded extends JSONView {};
+  static isolated = class Isolated extends JSONView {};
+  // The edit template is meant to be read-only, this field card is not mutable
+  static edit = class Edit extends JSONView {};
+}
+
+const messageStyle = {
+  boxelMessageAvatarSize: '2.5rem',
+  boxelMessageMetaHeight: '1.25rem',
+  boxelMessageGap: 'var(--boxel-sp)',
+  boxelMessageMarginLeft:
+    'calc( var(--boxel-message-avatar-size) + var(--boxel-message-gap) )',
+};
+
+class RoomMemberView extends Component<typeof RoomMemberCard> {
+  <template>
+    <div>
+      User ID:
+      {{@model.userId}}
+    </div>
+    <div>
+      Name:
+      {{@model.displayName}}
+    </div>
+  </template>
+}
+
+class RoomMemberCard extends Card {
+  @field userId = contains(StringCard);
+  @field displayName = contains(StringCard);
+  static embedded = class Embedded extends RoomMemberView {};
+  static isolated = class Isolated extends RoomMemberView {};
+  // The edit template is meant to be read-only, this field card is not mutable
```

Comment:
Seems like we this repeated pattern should make us think about what might be missing in the Card API. Do we need a way to express immutability?